### ISSUE TITLE
Move to epoch based training

### ISSUE THREAD

**Issue by tom9358** [tom9358/nllb-tryout#16]
Current training approach is a bit unconventional and hard to compare to other people's approaches. Let's reshape the training loop slightly so we have an epochs parameter.

Our training set keeps growing (slowly), so that makes the comparison between versions of the same model a bit different, because keeping the number of epochs constant means seeing more data points in total, whereas keeping the current settings - the training steps - constant, means seeing the same total number of data points (but as the training set grows, it means seeing each data point less often). I don't think that matters though. Still I don't think that matters much and using epochs should just make life easier.

Note to self - currently I am randomly sampling the data at each training step. Not only is that a bit cpu intensive, it also adds an unnecessary random factor to training. Probably better to pre-shuffle the entire training set at each epoch, and then sequentially go through it and train the model that way.

**Comment by tom9358** [comment]
I sped it up. tokenizer is now only called 2L times for L languages, and I did some optimizations of the variation adding functions. There are more things that can be sped up, but I don't think it's worth it at this moment. Although - possibly it would be smart to move the entire pretokenized epoch to device in one go, rather than per batch?? That's the last thing I'll look at, and then I'll close this issue.
